pennylane/devices/tests/test_properties.py

```diff
@@ -52,3 +87,144 @@ def test_no_0_shots(self, device_kwargs):
+            if TF_SUPPORT:
+                x = tf.Variable(0.1)
+                with tf.GradientTape() as tape:
+                    res = qnode(x)
+                    tape.gradient(res, [x])
+            else:
+                pytest.skip("Cannot import tensorflow.")
+
+        if interface == "autograd":
+            x = pnp.array(0.1, requires_grad=True)
+            g = qml.grad(qnode)
+            g(x)
+
+    def test_provides_jacobian(self, device_kwargs):
+        """Test that the device computes the jacobian."""
+        device_kwargs["wires"] = 1
+        dev = qml.device(**device_kwargs)
+        cap = dev.capabilities()
+
+        if "provides_jacobian" not in cap:
+            pytest.skip("No provides_jacobian capability specified by device.")
+
+        qnode = qml.QNode(qfunc_no_input, dev)
+
+        assert cap["provides_jacobian"] == hasattr(qnode, "jacobian")
+
+    def test_supports_tensor_observables(self, device_kwargs):
+        """Tests that the device reports correctly whether it supports tensor observables."""
+        device_kwargs["wires"] = 2
+        dev = qml.device(**device_kwargs)
+        cap = dev.capabilities()
+
+        if "supports_tensor_observables" not in cap:
+            pytest.skip("No supports_tensor_observables capability specified by device.")
+
+        qnode = qml.QNode(qfunc_tensor_obs, dev)
+
+        if cap["supports_tensor_observables"]:
+            qnode()
+        else:
+            with pytest.raises(qml.QuantumFunctionError):
+                qnode()
+
+    def test_reversible_diff(self, device_kwargs):
+        """Tests that the device reports correctly whether it supports reversible differentiation."""
+        device_kwargs["wires"] = 1
+        dev = qml.device(**device_kwargs)
+        cap = dev.capabilities()
+
+        if "supports_reversible_diff" not in cap:
+            pytest.skip("No supports_reversible_diff capability specified by device.")
+
+        if cap["supports_reversible_diff"]:
+            qfunc = qfunc_with_scalar_input(model=cap["model"])
+            qnode = qml.QNode(qfunc, dev, diff_method="reversible")
+            g = qml.grad(qnode)
+            g(0.1)
+        # no need to check else statement, since the reversible qnode creation fails in that case by default
+
+    def test_returns_state(self, device_kwargs):
+        """Tests that the device reports correctly whether it supports reversible differentiation."""
+        device_kwargs["wires"] = 1
+        dev = qml.device(**device_kwargs)
+        cap = dev.capabilities()
+
+        if "returns_state" not in cap:
+            pytest.skip("No returns_state capability specified by device.")
+
+        qnode = qml.QNode(qfunc_no_input, dev)
+        qnode()
+
+        if cap["returns_state"]:
+            dev.state  # pylint: disable = pointless-statement
+        else:
+            with pytest.raises(AttributeError):
+                dev.state  # pylint: disable = pointless-statement
```

Comment:
Some devices (e.g. `qiskit.aer`) have the `dev.state` attribute available, but it is equal to `None`. I would say that for such devices, `cap["returns_state"] = False` should be the case. However, I think this test would fail for such a device since it wouldn't raise an `AttributeError` :thinking:

If I understand the layout of the test suite correctly, this test would be applied to `qiskit.aer` in https://github.com/PennyLaneAI/plugin-test-matrix (I guess it would skip for now, until we define the capabilities for plugins).

Maybe we should have
```python
if cap["returns_state"]:
    assert dev.state is not None
else:
    try:
        state = dev.state
    except AttributeError:
        state = None

    assert state is None
```